lib/core/src/Cardano/Wallet/Registry.hs

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Cardano.Wallet.Registry
+    ( -- * Worker Registry
+      WorkerRegistry
+    , empty
+    , insert
+    , keys
+    , lookup
+    , remove
+
+      -- * Worker
+    , Worker
+    , MkWorker(..)
+    , newWorker
+    , workerThread
+    , workerId
+    , workerResource
+
+      -- * Context
+    , HasWorkerCtx (..)
+    ) where
+
+import Prelude hiding
+    ( log, lookup )
+
+import Cardano.BM.Trace
+    ( Trace, appendName )
+import Cardano.Wallet.Engine
+    ( HasLogger, logger )
+import Cardano.Wallet.Primitive.Types
+    ( WalletId )
+import Control.Concurrent
+    ( ThreadId, forkFinally, killThread )
+import Control.Concurrent.MVar
+    ( MVar
+    , modifyMVar_
+    , newEmptyMVar
+    , newMVar
+    , putMVar
+    , readMVar
+    , takeMVar
+    , tryPutMVar
+    )
+import Control.Exception
+    ( SomeException, finally )
+import Control.Monad.IO.Class
+    ( MonadIO, liftIO )
+import Data.Function
+    ( (&) )
+import Data.Generics.Internal.VL.Lens
+    ( (.~), (^.) )
+import Data.Generics.Labels
+    ()
+import Data.Generics.Product.Typed
+    ( HasType )
+import Data.Map.Strict
+    ( Map )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( toText )
+import GHC.Generics
+    ( Generic )
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+
+{-------------------------------------------------------------------------------
+                                Worker Context
+-------------------------------------------------------------------------------}
+
+-- | A class to link an existing context to a worker context.
+class HasType res (WorkerCtx ctx) => HasWorkerCtx res ctx where
+    type WorkerCtx ctx :: *
+    hoistResource :: res -> ctx -> WorkerCtx ctx
+
```

Comment:
We can go back to have a specialized type, but then we loose the benefits of being able to compose interfaces with only a subset of the wallet layer function (which means basically, going back and re-specialize the wallet).

Another option: have only function arguments instead of a `ctx`. However, this will become quite noisy and verbose. Maybe we could give it a try at some point :thinking: ... Having an abstract `ctx` makes it fairly easy to extend or change it without having to change a lot of functions depending on it.